### ISSUE TITLE
I have a weird Heisenbug

### ISSUE THREAD

**Issue by Midnighter** [opencobra/cobrapy#1180]
### Discussed in https://github.com/opencobra/cobrapy/discussions/1179

<div type='discussions-op-text'>

<sup>Originally posted by **akaviaLab** March 13, 2022</sup>
I've been trying to refactor gene.remove_from_model() and I have a test that fails when running the entire suite, but succeeds when running just that test.
Specifically, after my changes test_modify.py::test_rename_genes  fails when running the entire test suite, but succeeds when running everything.

If I change test_modify.py::test_rename_genes(model) to call test_modify.py::test_rename_genes(large_model) the test succeeds.

Do you guys have suggestions how to debug this?
I suspect that something is changing the model test_modify.py::test_rename_genes is using (model fixture), but I'm not sure how to identify where the change happens.</div>

**Comment by cdiener** [comment]
We probably also want to check that the reaction copying is not buggy. Looks like genes may not be copied as intended.

Edit: I get it now. It's like @matthiaskoenig said. Change the scope in the small_model fixture to function or better use the `model` fixture instead. Not sure why the small_model fixture was defined with session scope... Or are you using the model fixture and still seeing this?

**Comment by cdiener** [comment]
I checked and model copying seems to skip the GPR. This might lead to the observed behavior.

```python3
Python 3.9.10 (main, Feb 22 2022, 13:54:07) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.1.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from cobra.test import create_test_model

In [2]: mod = create_test_model("textbook")

In [3]: mod.reactions[0].gpr
Out[3]: cobra.core.gene.GPR('b0351 or b1241')

In [4]: id(mod.reactions[0].gpr)
Out[4]: 140284508959264

In [5]: mod2 = mod.copy()

In [6]: id(mod2.reactions[0].gpr)
Out[6]: 140284507356464

# works for reactions

In [9]: mod.reactions[0]
Out[9]: <Reaction ACALD at 0x7f96884c5ee0>

In [10]: mod2.reactions[0]
Out[10]: <Reaction ACALD at 0x7f96883a7c40>
```

**Comment by cdiener** [comment]
I checked and there seems to be an issue with copying as well. The expression body is copied by reference which is probably problematic.

```
Python 3.9.10 (main, Feb 22 2022, 13:54:07) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.1.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from cobra.test import create_test_model

In [2]: mod = create_test_model("textbook")

In [3]: mod2 = mod.copy()

In [4]: id(mod.reactions[0].gpr.body) == id(mod2.reactions[0].gpr.body)
Out[4]: True
```

I'll send in a PR for this (potentially unrelated) issue.

**Comment by akaviaLab** [comment]
I'm attaching a small python script that recapitulates the error on devel.
It seems  like there is some problem with Model.copy(), but I'm not sure. The ending is .txt so github won't complain. You can download it and then rename it - it is still a .py file.

[scratch_20.py.txt](https://github.com/opencobra/cobrapy/files/8240607/scratch_20.py.txt)

**Comment by akaviaLab** [comment]
The GPR is copied in reaction.copy() and model.copy(), so they should both be fixed. I think if you're doing that, it would be better to remove manual modifications of reaction._genes and use reaction._update_genes_from_gpr().  Let me know if you want me to do this fix. Thank you.
And I attached my test case without seeing you had already found the same thing. Sorry.

**Comment by cdiener** [comment]
It wasn't copied correctly on the current/previous devel at least (copy was to shallow). Or did you mean you already fixed it in your branch?

**Comment by akaviaLab** [comment]
Sorry, I wasn't clear.
It was copied shallowly. 

I think I fixed reaction.copy() and model.copy() on my branch right now, but for neatness, it seems to me like no other function should access reaction._genes, and use reaction._update_genes_from_gpr(). I've been working on that, but that's harder, especially with context. 

I can post the branch I've got if you want, but it isn't done. Something needs to happen with model.add_reactions() and model.remove_reactions() that are using reaction._genes, but should use gpr. Context is confusing to me.

**Comment by cdiener** [comment]
Yeah, I think they need to stay separate to some degree because SBML allows to have genes associated with a reaction even if there is no gene-reaction-rule and we need to stay compliant with SBML. So we need it to remain possible to add genes to a reaction without a GPR...

**Comment by akaviaLab** [comment]
I made additional changes. I didn't actually make your changes, but I rebased on them. @cdiener can you look at #1183 and see if that should be merged or not?

**Comment by cdiener** [comment]
The Heisenbug was due to the GPR being copied only by reference. This was fixed in #1181. So closing this here. A related issue is discussed in #1185.
